fdmi/plugins/m0scheduler

```diff
@@ -17,128 +17,277 @@
+# tool, forks m0sched program and waits for records delivered by it.
+# For now it is just a demo that prints FDMI records to console, but
+# later it can perform records de-dup and replication coordination by
+# means of set of the replication python programs that perform actual
+# data move.
+
 import subprocess
 import signal
-from socket import gethostname
 import shlex
 import sys
 import json
-import ast
 import codecs
 
-kv_records = [] #global kv cache
+cluster_info = dict()  # global cluster info
+kv_records = []        # global kv cache
+processes = []         # global list if m0sched processes
+
 
 def str_decode(encoded):
-	if encoded == "0" :
-		return encoded;
-	#return codecs.decode(encoded, 'hex').decode('utf-8')
-	return codecs.decode(encoded, 'hex')
+    if encoded is None:
+        return None
+    if encoded == '0':
+        return encoded
+    # return codecs.decode(encoded, 'hex').decode('utf-8')
+    return codecs.decode(encoded, 'hex')
+
 
 def json_parser(kv_record):
-	kvs = json.loads(kv_record)
-	kv = {'fid': str_decode(kvs["fid"]), 'cr_key': str_decode(kvs["cr_key"]), 'cr_val': str_decode(kvs["cr_val"])}
-	#kv = {'fid': kvs["fid"], 'cr_key': kvs["cr_key"], 'cr_val': kvs["cr_val"]}
-	flag =  True
-
-	#check dedup records
-	for kv_i in kv_records:
-		if kv['cr_key'] == kv_i['cr_key'] and kv['cr_val'] != "0":
-			#print(kv_records)
-			print("dedup: ", kv_i, kv)
-			flag = False
-			break
-
-	#if  record is not dedupable, add to cache
-	if flag == True and kv['cr_val'] != "0":
-		print("new kv: ", kv)
-		kv_records.append(kv)
-
-#handle SIGNT and SIGTERM
+    kvs = json.loads(kv_record)
+    fid = str_decode(kvs.get('fid'))
+    if fid is None:
+        print('Invalid fid in kv_record {}'.format(kv_record), file=sys.stderr)
+        return
+    cr_key = str_decode(kvs.get('cr_key'))
+    if cr_key is None:
+        print('Invalid cr_key in kv_record {}'.format(kv_record), file=sys.stderr)
+        return
+    cr_val = str_decode(kvs.get('cr_val'))
+    if cr_val is None:
+        print('Invalid cr_val in kv_record {}'.format(kv_record), file=sys.stderr)
+        return
+    kv = {
+        'fid': fid,
+        'cr_key': cr_key,
+        'cr_val': cr_val
+    }
+    # kv = {'fid': kvs["fid"], 'cr_key': kvs["cr_key"], 'cr_val': kvs["cr_val"]}
+    flag = True
+
+    # check dedup records
+    for kv_i in kv_records:
+        if kv['cr_key'] == kv_i['cr_key'] and kv['cr_val'] != "0":
+            # print(kv_records)
+            print('dedup: {} {}'.format(kv_i, kv), file=sys.stderr)
+            flag = False
+            break
+
+    # if record is not dedupable, add to cache
+    if flag and kv['cr_val'] != "0":
+        print('new kv: {}'.format(kv), file=sys.stderr)
+        kv_records.append(kv)
+
+
+# handle SIGNT and SIGTERM
 def signal_handler(sig, frame):
-	print("CTRL + C pressed")
-	process.send_signal(sig)
-	process.wait()
-	#print(kv_records)
-	sys.exit(0)
+    print('CTRL + C pressed', file=sys.stderr)
```

Comment:
Do we need:
global processes
at the beginning of this function?

Reply:
Dont recollect why we needed multiple processes, it was just one. @max-seagate  do you remember? If not required it can just be once process.

Reply:
The global was needed by my python version in order to mark the modification of the global var.

The list of the process is also an elegant way to handle warnings about accessing possibly not initialized (None) process. And yes, sometime in the future we may want to have more than one.

Reply:
This is just a sample python wrapper, don't see it launching multiple processes, even for prod use case.

Reply:
This is a clean way to do it... should I remove the list just because there is no plan to have more than 1 process?

Reply:
Nope did not suggest that :) We should do what is correct. Will this work for single process ? Or am I missing something here?
```
# define global process placeholder
_global_fdmi_proc = None
...
# create proc and set global var
...
# in current method use
if _global_fdmi_proc is not None:
   _global_fdmi_proc.send_signal(sig)
   _global_fdmi_proc.wait()
sys.exit(0)
```

Reply:
It will work but since the init is done conditionally it was showing me a warning that possible access to None is found.